torch/distributed/pipelining/schedules.py

```diff
@@ -603,6 +603,135 @@ def _step_microbatches(
+- prefetching can be as simple as moving an unshard 'backward' as many steps as possible without crossing a
+    different active layer, maybe?
+
+How should this code be structured?
+- send/recv is a global transform, we modify more than one rank at the same time.
+
+    Bad idea:
+    - its easiest to pick either 'sender' or 'receiever' as the source of truth, and then add the matching comm.
+    (e.g. if we prefer senders, we can iterate through rank0's actions and for every F, we insert a send and then
+        also insert a recv on the recieving rank)
+    - this seems totally error prone, actually. if we iterated through all of rank0's actions we'd end up appending
+      a bunch of recvs on other ranks.
+    but then whenn we iterate through other ranks actions and process their sends,
+    we'd need to insert recvs in 'the right place' in rank0 rather than just appending.
+
+    Idea 2 - we lower one whole 'time step' at a time
+        - at time 0 we have up to R actions.  For each action there are some sends.
+        - put those in a queue and then for each one,
+            append the send to the output trace of that rank and then append a recv to the recving rank
+
+- should we structure the FSDP stuff as a totally separate pass?
+    - its probably distracting to have to consider send/recv ops when doing prefetching
+    - not sure if it matters whether to start prefetch or start send/recv first relative to each other?
+    - Prefetching decisions are more 'across time for one rank' than 'across rank at one time',
+    so it seems messy to mix the two togehter
+
+-  maybe easiest is, do prefetch first per rank, then do p2p
+"""
+
+
+def _add_unshard_reshard(
+    compute_actions: Dict[int, List[Optional[_Action]]],
+    max_active_layers: int = 3,
+) -> Dict[int, List[Optional[_Action]]]:
+    """
+    Adds unshard/reshard actions to the schedule.
+
+    we abandon the "timestep lock"  during lowering
+
+    max_active_layers controls how many prefetches we allow. It should be measured in mb and tuneable but in practice
+    3 layers is probably the thing we want?  (to account for having one f and one b active, and something else prefetching?)
+    """
+
+    def next_stage_indices(next_actions: List[Optional[_Action]]) -> List[int]:
+        """Remove duplicates (same layer different microbatch) and tell me the order i'll encounter the stages in"""
+        seen: Set[int] = set()
+        ret: List[int] = []
+
+        # this is actually not the right algorithm, i think i should allow duplicates in ret, just remove consecutive
+        # instances of the same layer? hmm. that's not quite right either.
+        for a in next_actions:
+            if a is not None and a.stage_index not in seen:
+                seen.add(a.stage_index)
+                ret.append(a.stage_index)
+        return ret
+
+    def _per_rank_unshard(rank):
+        rank_actions = compute_actions[rank]
+        active_layers: Set[int] = set()
```

Comment:
What's the definition of `layer` here? As like a real "layer" in a model or something else? Reads like a concrete thing.

Reply:
yea i was being imprecise. i changed them all to 'stage'/'stages'